python/pdo/contract/replication.py

```diff
@@ -46,68 +43,78 @@
 __stop_service__ = False
 
 # -----------------------------------------------------------------
-def start_replication_service():
-
-    # start the manager
-    __replication_manager_executor__.submit(__replication_manager__)
+def start_replication_service(config = {}) :
+    # make sure the service is only started once
+    global replication_task_queue
+    with replication_thread_lock :
+        if replication_task_queue :
+            return
+        else :
+            replication_task_queue = queue.Queue()
+
+    global number_of_service_threads
+    number_of_service_threads = config.get("replication_worker_threads", number_of_service_threads)
+
+    logger.debug('start replication manager threads')
+    for i in range(config.get("replication_service_threads", 1)) :
+        thread = threading.Thread(target=__replication_manager__)
+        thread.daemon = True
+        thread.start()
+
+        with replication_thread_lock :
+            replication_threads.append(thread)
 
 # -----------------------------------------------------------------
-def stop_replication_service():
+def stop_replication_service() :
+    logger.debug("stop replication service")
 
     global __stop_service__
     __stop_service__ = True
 
-    #shutdown replication executor
-    __replication_manager_executor__.shutdown(wait=True)
+    for thread in replication_threads[:] :
+        thread.join(timeout=5.0)
```

Comment:
One behavior that I think we had in the current main (before this PR) is that stop_replicaation_service() call waits for the worker threads to finish processing their queues, they don't get aborted half-way. I think the new definition of stop_replication_service() continues to exhibit this earlier behaviour. 

Although what would perhaps be nice is an abort_replication_service() API that ideally should be called when we press "Ctrl + C" or when an Exception occurs anywhere within the stack, Of course, I am not sure how to distinguish the two at_exit scenarios to make the right call.

Reply:
this gives the threads five seconds to finish. that should be more than enough time. in the past, a hung thread would cause the whole process to hang.